includes/classes/SupportMonitor/DBQueryMonitor.php

```diff
@@ -0,0 +1,173 @@
+<?php
+/**
+ * Database Queries Monitor. A submodule of Support Monitor to report heavy SQL queries executed on staging.
+ *
+ * @since  x.x
+ * @package 10up-experience
+ */
+
+namespace TenUpExperience\SupportMonitor;
+
+/**
+ * DBQueryMonitor class
+ */
+class DBQueryMonitor {
+
+	/**
+	 * The transient name
+	 *
+	 * This transient stores all the queries logged.
+	 */
+	const TRANSIENT_NAME = 'tenup_experience_db_queries';
+
+	/**
+	 * Setup module
+	 */
+	public function setup() {
+		$production_environment = Monitor::instance()->get_setting( 'production_environment' );
+		if ( 'no' === $production_environment || apply_filters( 'tenup_experience_log_heavy_queries', false ) ) {
+			add_filter( 'query', [ $this, 'maybe_log_query' ] );
+		}
+	}
+
+	/**
+	 * Conditionally log a query
+	 *
+	 * Get all potential heavy queries (CREATE, ALTER, etc.) and store it,
+	 * ignoring transients and options by default.
+	 *
+	 * @param string $query The SQL query.
+	 * @return string
+	 */
+	public function maybe_log_query( $query ) {
+		global $wpdb;
+
+		if ( ! preg_match( '/^\s*(create|alter|truncate|drop|insert|delete|update|replace)\s/i', $query ) ) {
+			return $query;
+		}
+
+		if ( false !== strpos( $query, 'transient_' ) ) {
+			return $query;
+		}
+
+		if ( false !== strpos( $query, $wpdb->options ) ) {
+			return $query;
+		}
+
+		if ( apply_filters( 'tenup_experience_log_query', true, $query ) ) {
+			$this->log_query( $query );
+		}
+
+		return $query;
+	}
+
+	/**
+	 * Get the logged queries. Also removes from the transient all queries logged for more than 7 days.
+	 *
+	 * @return array
+	 */
+	public function get_report() {
+		$date_limit = strtotime( '7 days ago' );
```

Comment:
@felipeelia Messages are sent twice/day. My concern with only removing data within the past 7 days is we're going to have a lot of duplicate information sent during that 7-day period. It might make sense to remove the data as soon as the report data has been generated so we're only working with fresh data.

Reply:
@bengreeley the scenario I had in mind was: a Sys Eng made a plugin upgrade on staging on Friday but the Web Eng responsible for checking if queries are a concern or not are only able to check it later, on Monday, for example. As messages are erased, wouldn't we lose those queries?

Reply:
The way Support Monitor works is the 10up Experience plugin sends messages to the Support Monitor dashboard 2x daily, which is then turned into the latest report. In the scenario you mention as far as upgrading on Friday but checking for queries on Monday, you're right -- this wouldn't really catch those queries because they would have been logged already. You might want to consider the retention and see what makes sense. If there's an ALTER statement and it shows up for 7 days, is that accomplishing anything, or would it make more sense to be able to view that information directly on the site that is storing the data and just get an alert in Support Monitor that you should view the log because an excessive ALTER statement was called.

Support Monitor's main purpose is to keep an eye on things that may be problematic - not necessarily to store a bunch of data that gets wiped out. I'd recommend to reframe the approach of sending a message to Support Monitor of the status (like a count of ALTERs or other items in the log) and then have a place in the 10up Experience plugin you could go to in the WP admin to view the log, since those could be quite large.

Reply:
It is now storing the queries on the site, so the only thing in Support Monitor is if heavy queries were detected or not.